### DISCUSSION TITLE
How to perform "INSERT OR IGNORE" with SQLAlchemy 2.0 ORM?

### DISCUSSION THREAD

**Discussion by smoochy** [sqlalchemy/sqlalchemy#9675] [Usage Questions] [answered]
Hi community,

I am very new to Python and even newer to SQLAlchemy. I wanted to freshly start and just deal with the ORM [after watching the 2.0 video on Youtube by Mike](https://youtu.be/1Va493SMTcY).

When I add data to my SQLite DB, some of the columns have the `nullable=False` set. It is possible, that I add data (like a date) which is already in the DB. I simply want to ignore that data and not insert it into the DB.

I know that I could workaround with the following:

```python
stmt = (
            insert(Location)
            .values(
                provided_address=provided_address,
                formatted_address=formatted_address,
                geodata=geodata,
            )
            .prefix_with("OR IGNORE", dialect="sqlite")
        )
```

But I am not able to perform this with the plain ORM like presented in the same video, when adding it to the session with `session.add`. I also tried `session.merge`but it is (correctly) throwing the `sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed`.

Is there any way to realise an "INSERT OR IGNORE" with the ORM? I've [read a statement](https://github.com/sqlalchemy/sqlalchemy/issues/5374#issuecomment-638354961) from @zzzeek that it was not a priority. Is it possible by now, and do I just miss it?

Kind Regards.

**Comment by zzzeek** [comment, marked as answer]
the sqlite syntax we suppport is [insert..on conflict do update...or UPSERT](https://www.sqlite.org/lang_upsert.html) and we have that using a SQLite variant of `insert()` that's at https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#insert-on-conflict-upsert .

so if "id" were the column you're trying to be unique on:

```
from sqlalchemy.dialects.sqlite import insert
stmt = (
            insert(Location)
            .values(
                provided_address=provided_address,
                formatted_address=formatted_address,
                geodata=geodata,
            ).on_conflict_do_nothing(
                 index_elements=['id']
            )
        )
```

**Comment by smoochy** [comment]
Hi Mike,

thanks for the quick response. That will do as a workaround.

But my initial question (probably poorly expressed because I can't grasp all the different possibilities and labels of things currently).

Is it possible to do it with a syntax [like shown here in the upper part of the slide](https://youtu.be/1Va493SMTcY?t=6062)?

Lets say, I have a table `Date`, with a column `date` which is set to `unique=True`. That column has a row with the content `2023-04-19`

When I now want to do something like

```Python
with Session.begin() as session:
    session.add_all(
    [
        Date(date="2023-04-18"),
        Date(date="2023-04-19"),
        Date(date="2023-04-20"),
    ],
)
```

and `session.commit()` it, it will fail, because `2023-04-19` is already in the DB. How can this be solved?
I did not find a `session.add_on_conflict()` or similar method.

Kind Regards.

**Reply by CaselIT** [reply]
No, that's not available. Not all database support such feature, so it would be hard to generalize.
At the moment you will have to use the explicit insert on_conflict as mentioned in mike's answer

**Comment by smoochy** [comment]
Thanks for your reply and for confirming. Then I will use the method provided by Mike :)

Have a good one.

**Reply by zzzeek** [reply]
just as a note, the insert().on_conflict() thing is now presented as an ORM means of working also in the 2.0 series which you can see at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-queryguide-upsert 

for pure session.add() styles of working, which means we want to resolve the conflicts on the client side, there is a recipe that shows this general idea at https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject .   this is an old recipe that should still work but it probably could use an update.    It's also not organized to be efficient for many objects, though there are ways to do that client side without using an ON CONFLICT statement also.
